File: ios/TimberCruisingApp/Screens/MeasurementContinuationSheet.swift
// Post-measurement continuation prompt — what the cruiser sees after
// they hit Accept on a Quick Measure scan, before the cover dismisses
// them back to the hub.
//
// Cruiser feedback: "측정 이후에 워크플로우로 자연스럽게 넘어가게
// 해야지. dbh 재고 결과 나오고 땡이 아니라". Right — the natural
// continuation after a DBH on tree #7 is height on tree #7, then
// tree #8's DBH, etc. Dumping back to home meant 2 extra taps for
// each pivot. This sheet keeps the loop tight.
//
// Two presentation contexts:
//
//   • From the DBH cover's onAccept:
//       — "Measure height (tree #N)"  → Height scan, same tree
//       — "Next tree"                 → DBH scan, auto-incremented
//       — "Done"                      → back to hub
//
//   • From the Height cover's onAccept:
//       — "Next tree"                 → DBH scan, auto-incremented
//       — "Done"                      → back to hub
//
// Hidden when the cruiser already has BOTH dbh + height on that
// tree — at that point only "Next tree" / "Done" make sense.

import SwiftUI

public struct MeasurementContinuationSheet: View {

    public enum NextAction {
        case measureHeightSameTree
        case startNewTreeDiameter
        case done
    }

    public enum Origin {
        case afterDiameter
        case afterHeight

        var headlineSubject: String {
            switch self {
            case .afterDiameter: return "Diameter saved"
            case .afterHeight:   return "Height saved"
            }
        }
    }

    let origin: Origin
    let treeNumber: Int
    /// True when this tree already has a height reading (so we don't
    /// offer "Measure height" again for the same stem).
    let treeAlreadyHasHeight: Bool
    let onPick: (NextAction) -> Void

    @Environment(\.dismiss) private var dismiss

    public init(origin: Origin,
                treeNumber: Int,
                treeAlreadyHasHeight: Bool,
                onPick: @escaping (NextAction) -> Void) {
        self.origin = origin
        self.treeNumber = treeNumber
        self.treeAlreadyHasHeight = treeAlreadyHasHeight
        self.onPick = onPick
    }

    public var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: ForestixSpace.md) {
                    masthead
                    actions
                }
                .padding(.horizontal, ForestixSpace.md)
                .padding(.top, ForestixSpace.sm)
                .padding(.bottom, ForestixSpace.xl)
            }
            .background(ForestixPalette.canvas.ignoresSafeArea())
            .navigationTitle("Tree #\(treeNumber)")
            #if os(iOS)
            .navigationBarTitleDisplayMode(.inline)
            #endif
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Done") {
                        onPick(.done)
                        dismiss()
                    }
                }
            }
        }
        .presentationDetents([.medium, .large])
    }

    // MARK: - Masthead

    private var masthead: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(origin.headlineSubject)
                .font(ForestixType.title)
                .foregroundStyle(ForestixPalette.textPrimary)
            Text("What's next on tree #\(treeNumber)?")
                .font(ForestixType.body)
                .foregroundStyle(ForestixPalette.textSecondary)
        }
        .padding(.top, ForestixSpace.sm)
    }

    // MARK: - Actions

    private var actions: some View {
        VStack(spacing: ForestixSpace.sm) {
            // Same-tree height capture is the primary recommendation
            // after a DBH. Hidden when the tree already has height,
            // and not shown after a height scan.
            if origin == .afterDiameter, !treeAlreadyHasHeight {
                ActionCard(
                    title: "Measure height",
                    subtitle: "Stay on tree #\(treeNumber)",
                    systemImage: "arrow.up.and.down",
                    style: .primary
                ) {
                    onPick(.measureHeightSameTree)
                    dismiss()
                }
            }

            ActionCard(
                title: "Next tree",
                subtitle: "Start tree #\(treeNumber + 1) with a fresh diameter scan",
                systemImage: "plus.circle",
                style: origin == .afterHeight || treeAlreadyHasHeight
                    ? .primary : .secondary
            ) {
                onPick(.startNewTreeDiameter)
                dismiss()
            }

            ActionCard(
                title: "Back to hub",
                subtitle: "Stop measuring for now",
                systemImage: "house",
                style: .tertiary
            ) {
                onPick(.done)
                dismiss()
            }
        }
    }
}

// MARK: - Action card

private struct ActionCard: View {

    enum Style { case primary, secondary, tertiary }

    let title: String
    let subtitle: String
    let systemImage: String
    let style: Style
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: ForestixSpace.md) {
                ZStack {
                    RoundedRectangle(cornerRadius: ForestixRadius.control,
                                     style: .continuous)
                        .fill(glyphBackground)
                        .frame(width: 44, height: 44)
                    Image(systemName: systemImage)
                        .font(.system(size: 18, weight: .medium))
                        .foregroundStyle(glyphForeground)
                }
                VStack(alignment: .leading, spacing: 2) {
                    Text(title)
                        .font(ForestixType.bodyBold)
                        .foregroundStyle(titleColor)
                    Text(subtitle)
                        .font(ForestixType.caption)
                        .foregroundStyle(subtitleColor)
                        .lineLimit(2)
                }
                Spacer(minLength: 0)
                Image(systemName: "chevron.forward")
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundStyle(chevronColor)
            }
            .padding(ForestixSpace.md)
            .frame(maxWidth: .infinity)
            .background(cardBackground)
            .overlay(cardBorder)
        }
        .buttonStyle(.plain)
    }

    private var cardBackground: some View {
        RoundedRectangle(cornerRadius: ForestixRadius.card, style: .continuous)
            .fill(style == .primary
                  ? ForestixPalette.primary.opacity(0.10)
                  : ForestixPalette.surface)
    }

    private var cardBorder: some View {
        RoundedRectangle(cornerRadius: ForestixRadius.card, style: .continuous)
            .stroke(style == .primary
                    ? ForestixPalette.primary.opacity(0.35)
                    : ForestixPalette.divider,
                    lineWidth: style == .primary ? 1.0 : 0.5)
    }

    private var glyphBackground: Color {
        style == .primary
            ? ForestixPalette.primary
            : ForestixPalette.primaryMuted
    }

    private var glyphForeground: Color {
        style == .primary ? Color.white : ForestixPalette.primary
    }

    private var titleColor: Color {
        ForestixPalette.textPrimary
    }

    private var subtitleColor: Color {
        ForestixPalette.textSecondary
    }

    private var chevronColor: Color {
        style == .tertiary
            ? ForestixPalette.textTertiary
            : ForestixPalette.textSecondary
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/ui/screens/MeasurementContinuationSheet.kt
// Post-save continuation sheet — after a DBH or height is recorded, the
// next logical action (height on the same tree, or the next tree's
// diameter) is one tap away instead of dumping the cruiser back to the
// hub. Port of the iOS MeasurementContinuationSheet: bottom sheet with a
// "Tree #N" title bar, a masthead ("Diameter saved" / "Height saved" +
// "What's next on tree #N?"), and icon + subtitle action cards.

package com.hcjeong.forestix.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AddCircleOutline
import androidx.compose.material.icons.filled.ChevronRight
import androidx.compose.material.icons.filled.Height
import androidx.compose.material.icons.outlined.Home
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.hcjeong.forestix.ui.clickableNoRipple
import com.hcjeong.forestix.ui.theme.Forestix
import com.hcjeong.forestix.ui.theme.ForestixRadius
import com.hcjeong.forestix.ui.theme.ForestixSpace

/// What the cruiser chose from the continuation sheet.
enum class ContinuationAction { MEASURE_HEIGHT_SAME_TREE, START_NEW_TREE_DIAMETER, DONE }

/// Which measurement was just saved (drives the masthead + primary action).
enum class ContinuationOrigin { AFTER_DIAMETER, AFTER_HEIGHT }

private val ContinuationOrigin.headlineSubject: String
    get() = when (this) {
        ContinuationOrigin.AFTER_DIAMETER -> "Diameter saved"
        ContinuationOrigin.AFTER_HEIGHT -> "Height saved"
    }

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun MeasurementContinuationSheet(
    origin: ContinuationOrigin,
    treeNumber: Int,
    treeAlreadyHasHeight: Boolean,
    onAction: (ContinuationAction) -> Unit,
) {
    val type = Forestix.type
    val colors = Forestix.colors
    ModalBottomSheet(
        onDismissRequest = { onAction(ContinuationAction.DONE) },
        containerColor = colors.canvas,
    ) {
        Column(
            Modifier
                .fillMaxWidth()
                .padding(horizontal = ForestixSpace.md)
                .navigationBarsPadding()
                .padding(bottom = ForestixSpace.xl),
            verticalArrangement = Arrangement.spacedBy(ForestixSpace.md),
        ) {
            // Title bar — inline centred "Tree #N" + leading Done, the iOS
            // NavigationStack toolbar (cancellationAction) chrome.
            Box(Modifier.fillMaxWidth().height(44.dp)) {
                TextButton(
                    onClick = { onAction(ContinuationAction.DONE) },
                    modifier = Modifier.align(Alignment.CenterStart),
                ) { Text("Done") }
                Text(
                    "Tree #$treeNumber",
                    style = type.body.copy(fontSize = 17.sp, fontWeight = FontWeight.SemiBold),
                    color = colors.textPrimary,
                    modifier = Modifier.align(Alignment.Center),
                )
            }

            // Masthead — origin subject headline + the what's-next line.
            Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                Text(origin.headlineSubject, style = type.title, color = colors.textPrimary)
                Text(
                    "What's next on tree #$treeNumber?",
                    style = type.body,
                    color = colors.textSecondary,
                )
            }

            // Action cards.
            Column(verticalArrangement = Arrangement.spacedBy(ForestixSpace.sm)) {
                // Same-tree height capture is the primary recommendation
                // after a DBH. Hidden when the tree already has height,
                // and not shown after a height scan.
                if (origin == ContinuationOrigin.AFTER_DIAMETER && !treeAlreadyHasHeight) {
                    ContinuationActionCard(
                        title = "Measure height",
                        subtitle = "Stay on tree #$treeNumber",
                        icon = Icons.Filled.Height,
                        style = CardStyle.PRIMARY,
                    ) { onAction(ContinuationAction.MEASURE_HEIGHT_SAME_TREE) }
                }
                ContinuationActionCard(
                    title = "Next tree",
                    subtitle = "Start tree #${treeNumber + 1} with a fresh diameter scan",
                    icon = Icons.Filled.AddCircleOutline,
                    style = if (origin == ContinuationOrigin.AFTER_HEIGHT || treeAlreadyHasHeight) {
                        CardStyle.PRIMARY
                    } else {
                        CardStyle.SECONDARY
                    },
                ) { onAction(ContinuationAction.START_NEW_TREE_DIAMETER) }
                ContinuationActionCard(
                    title = "Back to hub",
                    subtitle = "Stop measuring for now",
                    icon = Icons.Outlined.Home,
                    style = CardStyle.TERTIARY,
                ) { onAction(ContinuationAction.DONE) }
            }
        }
    }
}

private enum class CardStyle { PRIMARY, SECONDARY, TERTIARY }

/// Port of the iOS ActionCard: 44 glyph tile + title/subtitle + chevron on
/// a card whose fill/border encode the recommendation strength.
@Composable
private fun ContinuationActionCard(
    title: String,
    subtitle: String,
    icon: ImageVector,
    style: CardStyle,
    onClick: () -> Unit,
) {
    val type = Forestix.type
    val colors = Forestix.colors
    val primary = style == CardStyle.PRIMARY
    Row(
        Modifier
            .fillMaxWidth()
            .clip(ForestixRadius.card)
            .background(if (primary) colors.primary.copy(alpha = 0.10f) else colors.surface)
            .border(
                width = if (primary) 1.dp else 0.5.dp,
                color = if (primary) colors.primary.copy(alpha = 0.35f) else colors.divider,
                shape = ForestixRadius.card,
            )
            .clickableNoRipple(onClick)
            .padding(ForestixSpace.md),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(ForestixSpace.md),
    ) {
        Box(
            Modifier
                .size(44.dp)
                .clip(ForestixRadius.control)
                .background(if (primary) colors.primary else colors.primaryMuted),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                icon,
                contentDescription = null,
                tint = if (primary) Color.White else colors.primary,
                modifier = Modifier.size(18.dp),
            )
        }
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Text(title, style = type.bodyBold, color = colors.textPrimary)
            Text(subtitle, style = type.caption, color = colors.textSecondary, maxLines = 2)
        }
        Icon(
            Icons.Filled.ChevronRight,
            contentDescription = null,
            tint = if (style == CardStyle.TERTIARY) colors.textTertiary else colors.textSecondary,
            modifier = Modifier.size(12.dp),
        )
    }
}
